### DISCUSSION TITLE
I'm not sure if I did the test correctly, but my test results are not ideal. Is there something wrong?

### DISCUSSION THREAD

**Discussion by newday11** [KyryloKuzyk/PrimeTween#37] [FAQ] [answered]
I'm not sure if I did the test correctly, but my test results are not ideal. Is there something wrong?
this is dotween pro and PrimeTween and IEnumerator
code ti is 

using System.Collections;
using System.Collections.Generic;
using DG.Tweening;
using UnityEngine;
using Tween = PrimeTween.Tween;

public class ddd : MonoBehaviour
{

    private WaitForSeconds rr;
    void Start()
    {
         rr = new WaitForSeconds(0.3f);;
    }

    // Update is called once per frame
    void Update()
    {
        if (Input.GetKeyDown(KeyCode.Q))
        {
            for (int i = 0; i < 100000; i++)
            {
                Tween.Delay(0.3f, () => { testscript();});
            }
        }
        if (Input.GetKeyDown(KeyCode.W))
        {
            for (int i = 0; i < 10000; i++)
            {
                DOVirtual.DelayedCall(0.3f, () => { testscript(); });
            }
        }
        if (Input.GetKeyDown(KeyCode.E))
        {
            for (int i = 0; i < 10000; i++)
            {
                StartCoroutine(test());
            }
        }
    }

    IEnumerator test()
    {
        yield return rr;
        testscript();
    }

    private void testscript()
    {
        int dd = 45;
    }
}

l am use mac os 
unity it is 2022.3.2f1c1
Below are my test results
PrimeTween:
first fps
![1067270973](https://github.com/KyryloKuzyk/PrimeTween/assets/93049155/6dc63a53-f8d5-452f-a3f5-ad7c73da07d0)
subsequent fps
![171722091](https://github.com/KyryloKuzyk/PrimeTween/assets/93049155/eb8fa5e3-daaa-4b11-bc37-29c9458250bd)

dotween pro
frist fps
![3465974413](https://github.com/KyryloKuzyk/PrimeTween/assets/93049155/e44d490a-2455-4b03-a37a-4bd474dd7e4d)
subsequent fps
![1454346979](https://github.com/KyryloKuzyk/PrimeTween/assets/93049155/18b09b6b-b90e-4962-a8af-5a8a3c44bd08)

IEnumerator
first fps
![58530264](https://github.com/KyryloKuzyk/PrimeTween/assets/93049155/cb73fadb-48df-4137-964c-799f7797016e)
subsequent fps
![3876541780](https://github.com/KyryloKuzyk/PrimeTween/assets/93049155/152cb668-d485-4197-996a-cadd5367cfd0)

As a result of this comparison, PrimeTween is very bad.
All 3 of them have gc on the first pin, and PrimeTween has the highest delay, so I feel like I did something wrong?

**Comment by KyryloKuzyk** [comment]
Hi,

Profiling in Editor typically doesn't give precise performance test results. Please check this [Benchmarks](https://github.com/KyryloKuzyk/PrimeTween/tree/main/Benchmarks) repo and run tests on a real device (preferably non-mobile because mobile phones have capped fps). See the "How to run the benchmarks?" section in [this article](https://github.com/KyryloKuzyk/PrimeTween/discussions/10) to ensure you're running PrimeTween with release performance.

Also, you're using an allocating version of Tween.Delay(). Please refer to this [documentation section](https://github.com/KyryloKuzyk/PrimeTween#zero-allocations-with-delegates) on how to prevent memory allocations in delegates.

**Reply by newday11** [reply]
help me

**Reply by KyryloKuzyk** [reply]
https://github.com/KyryloKuzyk/PrimeTween/discussions/37#discussioncomment-8118952

**Comment by newday11** [comment]
I modified it as requested to
Tween.Delay(this,0.3f, target => target.testscript());
But the result is still the same. There is no difference. What should I do?
the gc it is zero ,
but the time it is not good

**Reply by KyryloKuzyk** [reply]
@newday11 Profiling performance in the Editor with Profiler does not represent a real performance in release builds. Profiler is a tool to find bottlenecks and estimate the performance approximately. However, the real-world performance may differ by an order of magnitude, and this is what you’re experiencing. 

The correct way to measure performance precisely is to use Performance Testing Extensions AND measure in a real build (without Profiler). Please check the Benchmarks repo for an example.

Also, to measure a release performance, add the PRIME_TWEEN_DISABLE_ASSERTIONS define. This define will disable assertions and extra safety checks, so you can measure the actual release performance of the library.

**Reply by KyryloKuzyk** [reply, marked as answer]
@newday11 Oh, I just spotted an extra zero in your PrimeTween test, so the actual result should be 10 times quicker.

**Reply by newday11** [reply]
so thank you. but l am have other question；
I still have gc in the following two situations. I don’t know how to deal with it.
    public void EnemyInstantiatebuff(string ID, Enemymanager enemymanager)
    {
        Tween.Delay(this, 0.1f, aa => aa.AddEenemyBuff(ID, enemymanager));
    }
method data, I don't know how to write it, but this does generate gc

**Reply by KyryloKuzyk** [reply]
This is a hard one and there is no simple way to avoid GC. 

The code above generates GC because variables ID and enemymanager are captured in a closure. Creating a closure creates a new object instance that holds the captured variables.

I don't know the exact structure of your project, but let's suppose your code example is located at the Enemy.cs class. One way to eliminate the GC allocation is this one:
```csharp
public class Enemy : MonoBehaviour {
    // store the 'enemymanager' reference so it can be accessed in the Tween.Delay() callback via the Enemy reference 
    Enemymanager enemymanager;

    public void AddBuff() {
        // inline the EnemyInstantiatebuff() method to remove the capture of 'string ID' parameter  
        Tween.Delay(this, 0.1f, enemy => enemy.enemymanager.AddEenemyBuff(enemy, "buffID"));
    }
}
```

Another way is to simply let the allocation happen. If this code is not called every frame, but happens few times in a minute or so, it will not ruine your performance.

**Comment by newday11** [comment]
first fps 
![4152270425](https://github.com/KyryloKuzyk/PrimeTween/assets/93049155/847043f5-0d5a-4d6d-a55e-01bca757db9d)
subsequent fps
![2682337949](https://github.com/KyryloKuzyk/PrimeTween/assets/93049155/87bbec33-2a66-4768-95dd-7bf44c2d73d2)
